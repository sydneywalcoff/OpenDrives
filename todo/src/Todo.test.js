import React from 'react';
import Todos, { TodoList, TodoListItem, TodoForm, TodoHeader } from './Todo';

import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('todo Header', () => {
    it('renders without crashing', () => {
        render(<TodoHeader />);
    })

    it('matches snapshot', () => {
        const { asFragment } = render(<TodoHeader />);
        expect(asFragment()).toMatchSnapshot();
    })
})

describe('todo List Item', () => {
    const todo = { index: 1, value: "learn react", done: false };

    it('renders without crashing', () => {
        render(<TodoListItem ikey={todo.index} item={todo} index={todo.index} />)
    })

    it('matches snapshot', () => {
        const { asFragment } = render(<TodoListItem ikey={todo.index} item={todo} index={todo.index} />);
        expect(asFragment()).toMatchSnapshot();
    })

    it('renders to do list item with correct value', () => {
        const { getByTestId } = render(<TodoListItem ikey={todo.index} item={todo} index={todo.index} />);
        expect(getByTestId('todo-item')).toHaveTextContent("learn react");
    })

    it('renders status based on done property', () => {
        const { getByTestId } = render(<TodoListItem ikey={todo.index} item={todo} index={todo.index} />);
        expect(getByTestId('todo-status').innerHTML).toBe('[  ]')
    })

    it('changes className based on done property', () => {
        const { getByTestId } = render(<TodoListItem ikey={todo.index} item={todo} index={todo.index} />)
        expect(getByTestId('todo-status-class').className).toBe('undone')
    })

    it('renders delete button correctly', () => {
        const { getByTestId } = render(<TodoListItem ikey={todo.index} item={todo} index={todo.index} />);
        expect(getByTestId('todo-button').type).toBe('button');
        expect(getByTestId('todo-button').innerHTML).toBe('Ã—');
    })
})

describe('todo list', () => {
    const todoList = [{ index: 1, value: "learn react", done: false }, { index: 2, value: "brush teeth", done: false }, { index: 3, value: "walk dog", done: true }];

    it('renders todo list without crashing', () => {
        render(<TodoList items={todoList} />);
    })

    it('matches snpashot', () => {
        const { asFragment } = render(<TodoList items={todoList} />);
        expect(asFragment()).toMatchSnapshot();
    })

    it('renders the correct number of list items', () => {
        const { getByTestId } = render(<TodoList items={todoList} />);
        expect(getByTestId('todo-list').childElementCount).toBe(3);
    })
})

describe('todo form', () => {
    it('renders without crashing', () => {
        render(<TodoForm />);
    })

    it('matches snapshot', () => {
        const { asFragment } = render(<TodoForm />)
        expect(asFragment()).toMatchSnapshot()
    })

    it('renders input element correctly', () => {
        const { getByTestId } = render(<TodoForm />);
        expect(getByTestId('todo-input').placeholder).toBe("add a new todo...");
    })

    it('renders button correctly', () => {
        const { getByTestId } = render(<TodoForm />);
        expect(getByTestId('todo-form-button').type).toBe('submit');
        expect(getByTestId('todo-form-button')).toHaveTextContent('+');
    })
})

describe('create todo method', () => {
    const todoList = [
        {
            index: 1,
            value: "learn react",
            done: false
        }
    ]
    const newItemValue = 'walk dog'
    it('creates a new todo item', () => {
        const { getByTestId } = render(<Todos initItems={todoList} />);
        expect(getByTestId('todo-list').childElementCount).toBe(1);

        fireEvent.input(screen.getByTestId('todo-input'), {
            target: {
                value: newItemValue
            }
        })
        fireEvent.click(screen.getByTestId('todo-form-button'));
        expect(getByTestId('todo-list').childElementCount).toBe(2);
        expect(screen.getByTestId('todo-list')).toHaveTextContent(newItemValue);
    })
})

describe('complete todo method', () =>{
    const todoList = [{ index: 1, value: "learn react", done: false }, { index: 2, value: "brush teeth", done: false }, { index: 3, value: "walk dog", done: true }];

    it('changes done property to true', () => {
        const { getByText } = render(<Todos initItems={todoList} />)
        // check that todo Status is un-done
        const todoElText = getByText('learn react');
        // grab prev sibling to get to done status
        const todoStatusEl = todoElText.previousElementSibling;
        // check that done status is undone
        expect(todoStatusEl).toHaveTextContent("[ ]")
        // check that class name is 'undone' to apply undone styles
        expect(todoElText.className).toBe('undone')
        // click on the done span element
        fireEvent.click(todoStatusEl)
        // re-find the 'learn react' text
        const completedTodo = getByText('learn react');
        // get the previous element sibling to check the status
        const completedTodoStatus = completedTodo.previousElementSibling;
        // check that done status has been updated
        expect(completedTodoStatus).toHaveTextContent("[X]");
        // check that class name has been updated to 'done' to apply styles
        expect(completedTodo.className).toBe('done');
    })
})