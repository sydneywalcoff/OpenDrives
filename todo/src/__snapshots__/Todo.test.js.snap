// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo Header matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1
        class="Todo-header"
      >
        Todo list
      </h1>
    </div>
  </body>,
  "container": <div>
    <h1
      class="Todo-header"
    >
      Todo list
    </h1>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`todo List Item matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="Todo-item"
        data-testid="todo-item"
      >
        <span
          aria-hidden="true"
          class="icon"
          data-testid="todo-status"
        >
          [  ]
        </span>
        <div
          class="undone"
          data-testid="todo-status-class"
        >
          learn react
        </div>
        <button
          class="close"
          data-testid="todo-button"
          type="button"
        >
          ×
        </button>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="Todo-item"
      data-testid="todo-item"
    >
      <span
        aria-hidden="true"
        class="icon"
        data-testid="todo-status"
      >
        [  ]
      </span>
      <div
        class="undone"
        data-testid="todo-status-class"
      >
        learn react
      </div>
      <button
        class="close"
        data-testid="todo-button"
        type="button"
      >
        ×
      </button>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`todo list matches snpashot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="Todo-group"
        data-testid="todo-list"
      >
         
        <li
          class="Todo-item"
          data-testid="todo-item"
        >
          <span
            aria-hidden="true"
            class="icon"
            data-testid="todo-status"
          >
            [  ]
          </span>
          <div
            class="undone"
            data-testid="todo-status-class"
          >
            learn react
          </div>
          <button
            class="close"
            data-testid="todo-button"
            type="button"
          >
            ×
          </button>
        </li>
        <li
          class="Todo-item"
          data-testid="todo-item"
        >
          <span
            aria-hidden="true"
            class="icon"
            data-testid="todo-status"
          >
            [  ]
          </span>
          <div
            class="undone"
            data-testid="todo-status-class"
          >
            brush teeth
          </div>
          <button
            class="close"
            data-testid="todo-button"
            type="button"
          >
            ×
          </button>
        </li>
        <li
          class="Todo-item"
          data-testid="todo-item"
        >
          <span
            aria-hidden="true"
            class="icon"
            data-testid="todo-status"
          >
            [X]
          </span>
          <div
            class="done"
            data-testid="todo-status-class"
          >
            walk dog
          </div>
          <button
            class="close"
            data-testid="todo-button"
            type="button"
          >
            ×
          </button>
        </li>
         
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="Todo-group"
      data-testid="todo-list"
    >
       
      <li
        class="Todo-item"
        data-testid="todo-item"
      >
        <span
          aria-hidden="true"
          class="icon"
          data-testid="todo-status"
        >
          [  ]
        </span>
        <div
          class="undone"
          data-testid="todo-status-class"
        >
          learn react
        </div>
        <button
          class="close"
          data-testid="todo-button"
          type="button"
        >
          ×
        </button>
      </li>
      <li
        class="Todo-item"
        data-testid="todo-item"
      >
        <span
          aria-hidden="true"
          class="icon"
          data-testid="todo-status"
        >
          [  ]
        </span>
        <div
          class="undone"
          data-testid="todo-status-class"
        >
          brush teeth
        </div>
        <button
          class="close"
          data-testid="todo-button"
          type="button"
        >
          ×
        </button>
      </li>
      <li
        class="Todo-item"
        data-testid="todo-item"
      >
        <span
          aria-hidden="true"
          class="icon"
          data-testid="todo-status"
        >
          [X]
        </span>
        <div
          class="done"
          data-testid="todo-status-class"
        >
          walk dog
        </div>
        <button
          class="close"
          data-testid="todo-button"
          type="button"
        >
          ×
        </button>
      </li>
       
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
